### ISSUE TITLE
v1.80 test failures - are these expected?

### ISSUE THREAD

**Issue by c-w-m** [man-group/arctic#878]
#### Arctic Version
```
# Arctic version: 1.80.0
based on 12/22/20 clone of https://github.com/man-group/arctic.git
```
#### My Test System: 
* Ubuntu 16.04
* Python 3.6 environment
* Virtualenv 20.0.35 (.venv36)
* PyCharm 2020.3

#### Problem Description
* **close to ~300 test failures on first attempted test**
```python
python setup.py develop test
```

I suspect some of the dependencies I'm using do not match your development setup. I'll list my setup `requirements.txt` and `FIXME` notes which outline my attempted remedies.  

Any guidance on the setup or my approach to 'FIX' the issues would be greatly appreciated.

Thanks,
cm

----------------------------------------------------

## My Requirements.txt

```powershell
apipkg~=1.5
attrs~=20.2.0
certifi~=2020.12.5
cffi~=1.14.4
chardet~=3.0.4
contextlib2~=0.6.0.post1
coverage~=5.3.1
cryptography~=3.1.1
Cython~=0.29.21
decorator~=4.4.2
enum-compat~=0.0.3
enum34~=1.1.10
execnet~=1.7.1
Faker~=5.0.2
future~=0.18.2
idna~=2.10
inflection~=0.5.1
iniconfig~=1.1.1
ipython~=5.8.0
lz4~=3.1.1
mock~=4.0.3
mockextras~=1.0.2
nose~=1.3.7
numpy~=1.19.4
packaging~=20.8
pandas~=1.1.5
path~=15.0.1
pexpect~=4.8.0
pickleshare~=0.7.5
pip~=20.3.3
pluggy~=0.13.1
psutil~=5.8.0
ptyprocess~=0.6.0
py~=1.9.0
pycparser~=2.20
Pygments~=2.7.3
pymongo~=3.11.2
pyparsing~=2.4.7
pytest-cov~=2.10.1
pytest-server-fixtures~=1.7.0
pytest-timeout~=1.4.2
pytest-xdist~=1.26.1
pytest~=6.1.2
python-dateutil~=2.8.1
pytz~=2020.4
Quandl~=3.5.3
requests~=2.24.0
retry~=0.9.2
setuptools-git~=1.2
setuptools~=49.6.0
simplegeneric~=0.8.1
six~=1.15.0
toml~=0.10.2
traitlets~=4.3.3
tzlocal~=2.1
urllib3~=1.25.11
virtualenv~=20.0.35
wcwidth~=0.2.5
wheel~=0.36.2
zipp~=3.3.2
```

----------------------------------------------------

## My `FIXME` Notes
My attempts to remedy (or ignore) test errors:

> Note: the line numbers might differ from from your code base due to the comments and attempted FIX code that was added.

```
| Descriptions ....................................| Line Numbers | Notes
```

### FIXME: CM#001 - (str)
* tzlocal does not have a `zone` property , string cast
* look for benchmark impact
#### arctic
    - arctic/date/_mktz.py:                                  33 str cast
#### tests
    - tests/unit/date/test_mktz.py                            9 str cast

### FIXME: CM#002 - (xfail) - BUG? datastructure changed
* if the datastructure changed then update expected value by removing the missing field
#### xfail
* test_indexes(arctic):
#### tests
    - test_arctic.py                                         81-127 remove `u'ns': u'arctic.library',`

### FIXME: CM#003 - similar to (issue #420)
* date read from MongoDB does not come back with freq set (it is null)
#### new method
* add_idx_freq
#### tests
    - tests/util.py                                          77-91 add `add_idx_freq(idx, freq)`  (similarly for other files)
    - tests/integration/chuckstore/test_chunckstore.py       12 from tests.util import add_idx_freq
    - tests/integration/chuckstore/test_chunckstore.py      716-723 expose reads and add freq
    - tests/integration/chuckstore/test_chunckstore.py     1044-1046 expose reads and add freq
    - tests/integration/chunkstore/test_fixes.py             12 from tests.util import add_idx_freq
    - tests/integration/chunkstore/test_fixes.py            171-173 expose reads and add freq
    - tests/integration/store/test_pandas_store.py           16 from tests.util import add_idx_freq
    - tests/integration/store/test_pandas_store.py          858-904 expose reads and add freq
    - tests/integration/store/test_version_store.py          29

### FIXME: CM#004 - (xfail) - verify mock behavior of MongoDB transactions
* verify mock behavior of MongoDB transactions
#### xfail
* test_cleanup_orphaned_chunks(mongo_host, library, data, dry_run, fw_pointers_config)
* test_cleanup_orphaned_snapshots(mongo_host, library, data, dry_run, fw_pointers_config)
#### tests
    - tests/integration/scripts/test_arctic_fsck.py          32 xfail
    - tests/integration/scripts/test_arctic_fsck.py          50-56 expand for debug - no fix
    - tests/integration/scripts/test_arctic_fsck.py         144 xfail
    - tests/integration/scripts/test_arctic_fsck.py         174-175 expand for debug - no fix

### FIXME: CM#005 - (deprecate pandas.util.testing)
* replace `pandas.util.testing` to `pandas.testing`
#### arctic
    - arctic/tickstore/_util.py                               6
#### tests
    - tests/integration/chunckstore/tools/test_tools.py       6
    - tests/integration/chuckstore/test_chunckstore.py       10
    - tests/integration/chunkstore/test_fixes.py             10
    - tests/integration/chunckstore/test_utils.py             6
    - tests/integration/scripts/test_copy_data.py             3
    - tests/integration/store/test_bitemporal_store.py       10
    - tests/integration/store/test_metadata_store.py          9
    - tests/integration/store/test_pandas_store.py           13
    - tests/integration/store/test_version_store.py          14
    - tests/integration/store/test_version_store_audit.py     6
    - tests/integration/tickstore/test_toplevel.py            6
    - tests/integration/tickstore/test_ts_read.py             9
    - tests/integration/tickstore/test_ts_write.py            5
    - tests/integration/test_arctic.py                        6
    - tests/integration/test_copy_data.py                     3
    - tests/unit/chunkstore/test_date_chunker.py              7
    - tests/unit/serialization/test_numpy_arrays.py           4
    - tests/unit/store/test_bitemporal_store.py               4
    - tests/unit/tickstore/test_toplevel.py                  12
    - tests/unit/test_multi_index.py                          8

### FIXME: CM#006 - (xfail) - verify library.write and prune behavior
* verify library.write and prune behavior
#### xfail
* test_fix_broken_snapshot_references(library)
* test_keep_only_one_version(library)
* test_fix_broken_snapshot_references(library)
* test_keep_only_one_version(library)
#### tests
    - tests/integration/scripts/test_prune_versions.py        2 import pytest
    - tests/integration/scripts/test_prune_versions.py       54 xfail
    - tests/integration/scripts/test_prune_versions.py       57-67 expand for debug - no fix
    - tests/integration/scripts/test_prune_versions.py       70 xfail
    - tests/integration/scripts/test_prune_versions.py       74-84 expand for debug - no fix

### FIXME: CM#007 - (deprecated tostring)
* replace with tobytes() or str cast, `no change` for some code not yet revised
* look for benchmark impact
#### arctic
    - arctic/serialization/incremental.py                   223 tobytes
    - arctic/serialization/numpy_arrays.py                  119 tobytes
    - arctic/store/_ndarray_store.py                        449 tobytes
    - arctic/store/_ndarray_store.py                        600 tobytes
    - arctic/store/_ndarray_store.py                        657 tobytes
    - arctic/store/_pandas_ndarray_store.py                  58 tobytes
    - arctic/tickstore/tickstore.py                         707 tobytes
    - arctic/tickstore/tickstore.py                         718 tobytes
    - arctic/tickstore/tickstore.py                         727 tobytes
#### tests
    - tests/integration/store/test_ndarray_store_append.py  112 tobytes
    - tests/integration/store/test_ndarray_store_append.py  138 tobytes
    - tests/integration/store/test_ndarray_store_append.py  155 tobytes
    - tests/integration/store/test_ndarray_store_append.py  172 tobytes
    - tests/unit/serialization/test_incremental.py           53 str cast
    - tests/unit/serialization/test_incremental.py           73, 101, 136 tobytes

### FIXME: CM#008 - (expand cases for build_index_array)
* multiple cases need to be handled for `def _index_from_records(self, recarr)
* verify benchmark impact
#### arctic
arctic/serialization/numpy_records.py:                  95-107 add build_index_array

### FIXME: CM#009 - (replace assert_frame_equal with df.equals)
* replace assert_frame_equal with df.equals
#### tests
    - tests/integration/store/test_pickle_store.py:         802-826
    - tests/integration/store/test_pickle_store.py:         838, 843
    - tests/integration/store/test_pickle_store.py:         917-935
    - tests/integration/store/test_version_store.py        1162, 1165, 1183, 1186, 1234, 1241
    - tests/integration/store/test_version_store.py        1262, 1281, 1285, 1289, 1303, 1312
    - tests/integration/store/test_version_store.py        1365, 1382, 1403, 1426, 1446, 1465
    - tests/integration/store/test_version_store.py        1488
    - tests/integration/tickstore/test_toplevel.py           65, 84, 202, 205, 221
    - tests/integration/tickstore/test_ts_read.py           245-281

### FIXME: CM#010 - (deprecate numpy.testing.utils)
* replace numpy.testing.utils with numpy.testing
#### tests
    - tests/integration/tickstore/test_ts_read.py             9

### FIXME: CM#011 - (xfail) - BUG? read does not return Jan 1st
* read does not return January 1st
#### xfail
* test_should_return_data_when_date_range_falls_in_a_single_underlying_library(toplevel_tickstore, arctic)
* test_should_return_data_when_date_range_spans_libraries(toplevel_tickstore, arctic)
* test_should_write_top_level_with_list_of_dicts(arctic)
* test_should_write_top_level_with_correct_timezone(arctic)
#### tests
    - tests/integration/tickstore/test_toplevel.py           51 xfail
    - tests/integration/tickstore/test_toplevel.py           64 res read BUG?
    - tests/integration/tickstore/test_toplevel.py           68 xfail
    - tests/integration/tickstore/test_toplevel.py           76-80 expose variables for debug
    - tests/integration/tickstore/test_toplevel.py           82 res read BUG?
    - tests/integration/tickstore/test_toplevel.py          191 xfail
    - tests/integration/tickstore/test_toplevel.py          201 res read BUG?
    - tests/integration/tickstore/test_toplevel.py          208 xfail
    - tests/integration/tickstore/test_toplevel.py          220 res read BUG?

### FIXME: CM#012 - (xfail) - BUG? first value in CLOSED-CLOSED read is not returned
* correct expected value
#### xfail
* test_date_range(tickstore_lib)
* test_date_range_end_not_in_range(tickstore_lib)
* test_date_range_default_timezone(tickstore_lib, tz_name)
* def test_date_range_no_bounds(tickstore_lib)
#### tests
    tests/integration/tickstore/test_toplevel.py            232, 287, 312, 347 

### FIXME: CM#013 - (xfail) - BUG? losing_writer did not call `delete_version`
* losing_writer did not call `delete_version`
#### xfail
* test_ArcticTransaction_detects_concurrent_writes()
#### tests
    tests/unit/store/test_version_store_audit.py            179

### FIXME: CM#014 - (xfail) - (?query string format changed)
* the query string content matches with the exception of some added detail
#### xfail
* test_mongo_date_range_query()
#### tests
    tests/unit/tickstore/test_tickstore.py                   43

### FIXME: CM#015 - (xfail) - (?index string format changed)
* the index string content matches but there is some added detail
#### xfail
* test_tickstore_to_bucket_with_image()
* test_tickstore_pandas_to_bucket_image()
#### tests
    tests/unit/tickstore/test_tickstore.py                   99
    tests/unit/tickstore/test_tickstore.py                  169

### FIXME: CM#016 - (xfail) - (OverlappingDataException)
* toplevel_tickstore.add fails with OverlappingDataException
#### xfail
* test_should_add_underlying_library_where_another_library_exists_in_a_non_overlapping_daterange(toplevel_tickstore, arctic)
#### tests
    tests/integration/tickstore/test_toplevel.py            118

**Comment by bmoscon** [comment]
no, none of them are expected, and all need to be fixed

**Comment by dragonframework** [comment]
Do you want my changes on a new branch to decide if you want to merge them back into the master?

**Comment by bmoscon** [comment]
sure if you open a PR we'd probably take the fixes

**Comment by c-w-m** [comment]
Hi Bryant ( @bmoscon ),
I pushed the partial fix changes (leaving out the skip test changes).  My notes are in the /docs/history directory and most the code changes have a `#FIXME: CM...` comment to help find like items.

I've also added tox.ini to help setup env for dev/test, and benchmarking  Initially I had tox.ini use setup.py to determine the requirement packages but after looking into `.travis.yml` I see pip is used to update many items to that has also been added to the `tox.ini`.  Similarly, the benchmark env are manually setup in `.tox.ini` (but not run) so they are ready for benchmarking.  The asv install looks for the /asv submodule so the benchmark work can be easily controlled to a tagged version if needed.

I'm still kicking the tires on this project but I very much like what the team has put together ;-)

Hope this is helpful,
cm
